File: src/main/java/ui/BrowserFactory.java
package ui;

import com.microsoft.playwright.*;
import common.exceptions.MissingSystemPropertyException;
import common.SystemPropertyReader;

public class BrowserFactory {
    private static final ThreadLocal<Playwright> playwrightThread = new ThreadLocal<>();
    private static final ThreadLocal<BrowserType> browserTypeThread = new ThreadLocal<>();
    private static final ThreadLocal<Browser> browserThread = new ThreadLocal<>();
    private static final ThreadLocal<BrowserContext> browserContextThread = new ThreadLocal<>();
    private static final ThreadLocal<Page> pageThread = new ThreadLocal<>();

    public static synchronized Page getPage(Page page) throws MissingSystemPropertyException {
        if (page != null) {
            pageThread.set(page);
        } else if (playwrightThread.get() == null && page == null) {
            Playwright playwright = Playwright.create();
            playwrightThread.set(playwright);
            page = createPage(playwright);
            pageThread.set(page);
        }
        return pageThread.get();
    }

    public static synchronized Browser getBrowser() {
        // FIXME
        return browserThread.get();
    }

    public static synchronized void closePage() {
        Page page = pageThread.get();
        if (page != null) {
            page.close();
            pageThread.remove();
        }
    }

    public static synchronized void closePlaywright() {
        Playwright playwright = playwrightThread.get();
        if (playwright != null) {
            playwright.close();
            playwrightThread.remove();
        }
    }

    public static synchronized void closeBrowserContext() {
        BrowserContext browserContext = browserContextThread.get();
        if (browserContext != null) {
            browserContext.close();
            browserContextThread.remove();
        }
    }

    public static synchronized void closeBrowser() {
        Browser browser = browserThread.get();
        if (browser != null) {
            browser.close();
            browserThread.remove();
        }
    }

    private static synchronized Page createPage(Playwright playwright) throws MissingSystemPropertyException {
        BrowserType browserType = getBrowserType(playwright);
        Browser browser = browserType.launch();
        BrowserContext context = browser.newContext();;

        browserTypeThread.set(browserType);
        browserThread.set(browser);
        browserContextThread.set(context);
        return context.newPage();
    }

    private static synchronized BrowserType getBrowserType(Playwright playwright) throws MissingSystemPropertyException {
        String browserName = SystemPropertyReader.readSystemProperty("browserName");

        switch (browserName) {
            case "chromium":
                return playwright.chromium();
            case "webkit":
                return playwright.webkit();
            case "firefox":
                return playwright.firefox();
            default:
                throw new IllegalArgumentException("Invalid 'browserName' property: " + browserName);
        }
    }
}

File: src/main/kotlin/ui/BrowserFactory.kt
package ui

import com.microsoft.playwright.*
import common.SystemPropertyReader.readSystemProperty
import common.exceptions.MissingSystemPropertyException

object BrowserFactory {
    private val playwrightThread = ThreadLocal<Playwright?>()
    private val browserTypeThread = ThreadLocal<BrowserType>()
    private val browserThread = ThreadLocal<Browser>()
    private val browserContextThread = ThreadLocal<BrowserContext>()
    private val pageThread = ThreadLocal<Page?>()

    
    @Synchronized
    @Throws(MissingSystemPropertyException::class)
    fun getPage(page: Page?): Page? {
        var page = page
        if (page != null) {
            pageThread.set(page)
        } else if (playwrightThread.get() == null) {
            val playwright = Playwright.create()
            playwrightThread.set(playwright)
            page = createPage(playwright)
            pageThread.set(page)
        }
        return pageThread.get()
    }

    
    @get:Synchronized
    val browser: Browser
        get() =// FIXME
            browserThread.get()

    
    @Synchronized
    fun closePage() {
        val page = pageThread.get()
        if (page != null) {
            page.close()
            pageThread.remove()
        }
    }
    
    @Synchronized
    fun closePlaywright() {
        val playwright = playwrightThread.get()
        if (playwright != null) {
            playwright.close()
            playwrightThread.remove()
        }
    }
    
    @Synchronized
    fun closeBrowserContext() {
        val browserContext = browserContextThread.get()
        if (browserContext != null) {
            browserContext.close()
            browserContextThread.remove()
        }
    }

    @Synchronized
    fun closeBrowser() {
        val browser = browserThread.get()
        if (browser != null) {
            browser.close()
            browserThread.remove()
        }
    }

    @Synchronized
    @Throws(MissingSystemPropertyException::class)
    private fun createPage(playwright: Playwright): Page {
        val browserType = getBrowserType(playwright)
        val browser = browserType.launch()
        val context = browser.newContext()



        browserTypeThread.set(browserType)
        browserThread.set(browser)
        browserContextThread.set(context)
        return context.newPage()
    }

    @Synchronized
    @Throws(MissingSystemPropertyException::class)
    private fun getBrowserType(playwright: Playwright): BrowserType {
        val browserName = readSystemProperty("browserName")

        return when (browserName) {
            "chromium" -> playwright.chromium()
            "webkit" -> playwright.webkit()
            "firefox" -> playwright.firefox()
            else -> throw IllegalArgumentException("Invalid 'browserName' property: $browserName")
        }
    }
}
